src/ArrayValue.php

```diff
@@ -110,68 +119,75 @@ public function offsetExists($offset): bool;
 
     /**
      * @param int $offset
-     * @return mixed
+     * @phpstan-return TValue
      */
     public function offsetGet($offset);
 
     /**
      * @param int $offset
-     * @param mixed $value
-     * @return void
-     * @throws \BadMethodCallException For immutable types.
+     * @param TValue $value
+     * @throws BadMethodCallException For immutable types.
      */
     public function offsetSet($offset, $value): void;
 
     /**
      * @param int $offset
      * @return void
-     * @throws \BadMethodCallException For immutable types.
+     * @throws BadMethodCallException For immutable types.
      */
     public function offsetUnset($offset): void;
 
     // ArrayValue own
 
     /**
-     * @return ArrayValue
+     * @phpstan-param ArrayValue<TValue> $other
+     * @phpstan-return ArrayValue<TValue>
      */
     public function join(ArrayValue $other): ArrayValue;
 
     /**
-     * @return ArrayValue
+     * @phpstan-return ArrayValue<TValue>
      */
     public function slice(int $offset, int $length): ArrayValue;
 
     /**
-     * @return ArrayValue
+     * @phpstan-param ArrayValue<TValue> $replacement
+     * @phpstan-return ArrayValue<TValue>
      */
     public function splice(int $offset, int $length, ?ArrayValue $replacement = null): ArrayValue;
 
     /**
-     * @param callable|null $comparator function(mixed $valueA, mixed $valueB): int{-1, 0, 1}
-     * @return ArrayValue
+     * @phpstan-param ArrayValue<TValue> $other
+     * @param callable(TValue $valueA, TValue $valueB): int|null $comparator
+     * @phpstan-return ArrayValue<TValue>
```

Comment:
What if I have type `A` and `B` and a comparator that can to compare `A <=> B`?

Also `callable(TValue $valueA, TValue $valueB): int|null` seems like callable can return `int|null`

Reply:
I `A` and `B` does not share common parent class - you cannot have it in one `ArrayValue`